internal/pkg/aws/cloudwatchlogs/cloudwatchlogs.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package cloudwatchlogs contains utility functions for Cloudwatch Logs client.
+package cloudwatchlogs
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+)
+
+const (
+	// SleepDuration is the sleep time for making the next request for log events.
+	SleepDuration = 1 * time.Second
+)
+
+var (
+	fatalCodes   = []string{"FATA", "FATAL", "fatal", "ERR", "ERROR", "error"}
+	warningCodes = []string{"WARN", "warn", "WARNING", "warning"}
+)
+
+type cloudwatchlogsClient interface {
+	DescribeLogStreams(input *cloudwatchlogs.DescribeLogStreamsInput) (*cloudwatchlogs.DescribeLogStreamsOutput, error)
+	GetLogEvents(input *cloudwatchlogs.GetLogEventsInput) (*cloudwatchlogs.GetLogEventsOutput, error)
+}
+
+// Service wraps an AWS Cloudwatch Logs client.
+type Service struct {
+	logStreamName map[string]string // cache the latest updated log stream name for a certain log group
+	cwlogs        cloudwatchlogsClient
+}
+
+// New returns a Service configured against the input session.
+func New(s *session.Session) *Service {
+	return &Service{
+		cwlogs:        cloudwatchlogs.New(s),
+		logStreamName: make(map[string]string),
+	}
+}
+
+// logStreams returns the log stream name of the log stream with the latest log event in a log group.
+func (s *Service) logStreams(logGroupName string) (string, error) {
+	resp, err := s.cwlogs.DescribeLogStreams(&cloudwatchlogs.DescribeLogStreamsInput{
+		LogGroupName: aws.String(logGroupName),
+		Descending:   aws.Bool(true),
+		OrderBy:      aws.String("LastEventTime"),
+	})
+	if err != nil {
+		return "", fmt.Errorf("describe log streams of log group %s: %w", logGroupName, err)
+	}
+	if len(resp.LogStreams) == 0 {
+		return "", fmt.Errorf("no log stream found in log group %s", logGroupName)
+	}
+	return aws.StringValue(resp.LogStreams[0].LogStreamName), nil
+}
+
+// LogEvents returns an array of Cloudwatch Logs events.
+func (s *Service) LogEvents(logGroupName string, limit int64, startTime, endTime *int64, nextToken *string) ([]*Event, *string, error) {
```

Comment:
I've been trying for a while to use time.Time. However, the problem is if we use time.Time, since startTime and endTime are optional parameters, using *int64 we can easily pass them directly (and will not be used if they are not set). Using time.Time then we need to in the LogEvents function determine if they are set or not, which requires an extra two steps. What do you think?